### PR TITLE
feat(auth, emulator): implement useEmulator with jest + e2e testing

### PR DESCRIPTION
**Pull request by mikehardy**
### Description

Implement `firebase.auth().useEmulator`

I'm about to release a major version anyway so I included the removal of deprecated analytics methods

I intend to rebase on to master so these will stay as separate commits and will be picked up correctly for release

It should be reviewed commit by commit for clarity

Phone flow is not implemented yet - may get that in a follow-on commit here or in a follow-on PR. It is possible but you have to interact with the REST API on the emulator to do it.

### Related issues

#4461 
#4463 
#4510 
#4551 

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [x] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [x] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

This enabled me to turn on auth e2e testing again, and I implemented jest tests even though they were really difficult because of the native mocking need

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

File: packages/app/e2e/helpers.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+  return 'react-native-firebase-testing';
+};
+
+// Android requires a re-mapped hostname during testing sometimes
+exports.getE2eEmulatorHost = function getE2eEmulatorHost() {
+  // This appears to work from localhost even on android. Different execution context?
+  // if (platform === 'android') {
+  //   return '10.0.2.2';
+  // }
+  return 'localhost';
```
Comment (Salakar) [inline packages/app/e2e/helpers.js:11]: This works as localhost due to; https://github.com/invertase/react-native-firebase/pull/4552/files#diff-17651da2be070111f5b3f961a4e6fda957ed253e301e14f8a1da4a5ce95c44efR343-R351

We dod this on all emulator apis for RNFB so the users don't need to worry about it

File: packages/auth/e2e/auth.e2e.js [hunk 2]
```diff
@@ -15,7 +15,41 @@
...
+    // Need a way to create the user but have it disabled immediately
+    // If we had the functions emulator running as well, it would work:
+    // https://stackoverflow.com/questions/43140441/how-to-disable-user-accounts-on-firebase-authentication-automatically#comment73363725_43141780
+    // try {
+    //   const userCredential = await firebase
+    //     .auth()
+    //     .createUserWithEmailAndPassword(DISABLED_EMAIL, DISABLED_PASS);
+    // } catch (e) {
+    //   // they may already exist, that's fine
+    // }
```
Comment (Salakar) [inline packages/auth/e2e/auth.e2e.js:43]: Can hold off for now and hopefully in future something like this becomes available on the emulator 🤞

File: packages/auth/e2e/helpers.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
+/* eslint-disable no-console */
+const { getE2eTestProject, getE2eEmulatorHost } = require('../../app/e2e/helpers');
+const http = require('http');
+
+exports.getRandomPhoneNumber = function getRandomPhoneNumber() {
+  return '+593' + Utils.randString(9, '#19');
+};
+
+exports.clearAllUsers = async function clearAllUsers() {
```
Comment (Salakar) [inline packages/auth/e2e/helpers.js:9]: Off topic but curious if a disable user API could be reverse engineered of the Emulator Web UI requests as well, does the emulator web ui have an option to disable a user?

**Review by Salakar — APPROVED** [review]

File: packages/app/e2e/helpers.js (hunk 1 above)
Comment (mikehardy) [reply packages/app/e2e/helpers.js:11]: Yeah, I got that part from the functions emulator implementation and it works great. This is separate though. This helper function is returning a string without any intervening translation to a higher layer of helper function which constructs a raw URL out of it with no other translation, and only localhost works. It makes me think the e2e javascript is running in the context of jet on the host as opposed to the react-native JS executor in the emulator. I verified that you need 10.0.2.2 if you try it on the emulator (using `adb shell` and then `nc` in that shell to pretend I was a web client).

TL:DR; which you probably know: the e2e javascript is calling functions in the emulator but the regular javascript, is it executing in jet on the host?

File: packages/auth/e2e/helpers.js (hunk 3 above)
Comment (mikehardy) [reply packages/auth/e2e/helpers.js:9]: Oh it does! This might solve the problem. I will check if it's possible

Comment (mikehardy) [reply packages/auth/e2e/helpers.js:9]: Looks like a POST to the standard APIs works against the emulator! The docs were not clear on that, they have an "emulator specific" setting but don't mention if the non-specific APIs work, apparently they do:

```
scheme | http
-- | --
host | localhost:9099
filename | /identitytoolkit.googleapis.com/v1/accounts:update

{"disableUser":true,"localId":"tUkcA81QgKn0l6n1CXl3EFW1kR1x"}
```

Needs the same 'Authorization: Bearer owner' trick from account purge REST API call, but I already have the infrastructure to do REST API calls to the emulator so this is easy peasy now.

Comment (Salakar) [reply packages/auth/e2e/helpers.js:9]: Nice!

File: packages/app/e2e/helpers.js (hunk 1 above)
Comment (Salakar) [reply packages/app/e2e/helpers.js:11]: Oh ye sorry I misread this, e2e JS does run in Node.js on the host machine and not the emulator - so this makes sense.

Comment (mikehardy) [reply packages/app/e2e/helpers.js:11]: perfect, I will update that comment as "where things run" is vital to know, especially since we (as mentioned) do the translation to 10.0.2.2 in other places.

Side note: I do the equivalent of rebase -i for these PRs so the commits can stay separate, I've taken care that they'll even survive (with correct success/fail) if applied partially but in sequence so git bisect will work (something we learned to be careful about in AnkiDroid). Not the standard I would hold most contributors too but I think these commits serve best if merged in separated, similar to my iosv7/androidv26 forward port
